cli/archive.go

```diff
@@ -0,0 +1,114 @@
+package cli
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/fs"
+	"math"
+	"os"
+	"path/filepath"
+
+	"go.viam.com/utils"
+	"golang.org/x/exp/maps"
+)
+
+// getArchiveFilePaths traverses the provided rootpaths recursively,
+// collecting the file paths of all regular files and returns them in a slice.
+// This list of paths should be passed to createArchive.
+func getArchiveFilePaths(rootpaths []string) ([]string, error) {
+	files := map[string]bool{}
+	for _, pathRoot := range rootpaths {
+		err := filepath.WalkDir(filepath.Clean(pathRoot), func(path string, info fs.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.Type().IsRegular() {
+				files[path] = true
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return maps.Keys(files), nil
+}
+
+// createArchive compresses and archives the provided file paths into a tar.gz format,
+// writing the resulting binary data to the supplied "buf" writer.
+// If "stdout" is provided, the function outputs compression progress information.
+func createArchive(files []string, buf io.Writer, stdout *io.Writer) error {
+	// Create new Writers for gzip and tar
+	// These writers are chained. Writing to the tar writer will
+	// write to the gzip writer which in turn will write to
+	// the "buf" writer
+	gw := gzip.NewWriter(buf)
+	defer utils.UncheckedErrorFunc(gw.Close)
```

Comment:
I don't know what this does exactly but FYI docs say 'never use this for closing writers'

https://github.com/viamrobotics/goutils/blob/23898314dc98153baf6f45dd3ec142cd92200639/error.go#L69